model/task/task.go

```diff
@@ -2513,82 +2515,84 @@ func (t *Task) Insert() error {
+// expects that each one is going to be archived and progressed to the next execution.
+// For execution tasks in display tasks, it will properly account for archiving
+// only tasks that should be if failed.
 func ArchiveMany(tasks []Task) error {
-	if len(tasks) == 0 {
-		return nil
-	}
-	existingTasksMap := map[string]bool{}
-	for _, t := range tasks {
-		existingTasksMap[t.Id] = true
-	}
-	additionalTasks := []string{}
+	allTaskIds := []string{}          // Contains all tasks and display tasks IDs
+	execTaskIds := []string{}         // Contains all exec tasks IDs
+	toUpdateExecTaskIds := []string{} // Contains all exec tasks IDs that should update and have new execution
+	archivedTasks := []interface{}{}  // Contains all archived tasks (task, display, and execution). Created by Task.makeArchivedTask()
+
 	for _, t := range tasks {
-		// For any display tasks here, make sure that we also archive all their execution tasks.
-		// Consider each execution tasks individually, so we don't query a task we already passed in.
-		for _, et := range t.ExecutionTasks {
-			if !existingTasksMap[et] {
-				additionalTasks = append(additionalTasks, et)
-				existingTasksMap[et] = true
+		allTaskIds = append(allTaskIds, t.Id)
+		archivedTasks = append(archivedTasks, t.makeArchivedTask())
+		if t.DisplayOnly && len(t.ExecutionTasks) > 0 {
+			var eTasks []Task
+			var err error
+
+			if t.ResetFailedWhenFinished {
+				eTasks, err = Find(FailedTasksByIds(t.ExecutionTasks))
+			} else {
+				eTasks, err = FindAll(db.Query(ByIds(t.ExecutionTasks)))
+			}
+
+			if err != nil {
+				return errors.Wrapf(err, "finding execution tasks for display task '%s'", t.Id)
+			}
+			execTaskIds = append(execTaskIds, t.ExecutionTasks...)
+			for _, et := range eTasks {
+				archivedTasks = append(archivedTasks, et.makeArchivedTask())
+				toUpdateExecTaskIds = append(toUpdateExecTaskIds, et.Id)
 			}
 		}
 	}
-	if len(additionalTasks) > 0 {
-		toAdd, err := FindAll(db.Query(ByIds(additionalTasks)))
-		if err != nil {
-			return errors.Wrap(err, "finding execution tasks")
-		}
-		tasks = append(tasks, toAdd...)
-	}
 
-	archived := []interface{}{}
-	taskIds := []string{}
-	for _, t := range tasks {
-		archived = append(archived, *t.makeArchivedTask())
-		taskIds = append(taskIds, t.Id)
-	}
-	grip.DebugWhen(len(utility.UniqueStrings(taskIds)) != len(taskIds), message.Fields{
+	grip.DebugWhen(len(utility.UniqueStrings(allTaskIds)) != len(allTaskIds), message.Fields{
 		"ticket":           "EVG-17261",
 		"message":          "archiving same task multiple times",
-		"tasks_to_archive": taskIds,
+		"tasks_to_archive": allTaskIds,
 	})
 
+	return archiveAll(allTaskIds, execTaskIds, toUpdateExecTaskIds, archivedTasks)
+}
+
+// archiveAll takes in:
+// - taskIds               : All tasks and display tasks IDs
+// - execTaskIds           : All execution task IDs
+// - toUpdateExecTaskIds   : All execution task IDs for execution tasks that will be archived/restarted
```

Comment:
Sorry I think this is a repeat comment but I can't find it if you answered it -- can this be called toRestartExecTaskIds instead? Or something like that.  ToUpdate is confusing since we also update the execTaskIds.

Reply:
Yes! That sounds much cleaner, I will rename